src/C4EnergyBars.cpp

```diff
@@ -0,0 +1,586 @@
+		hashCombine(result, std::hash<uint32_t>{}(gfx.second.amount));
+		hashCombine(result, std::hash<uint32_t>{}(gfx.second.scale));
+	}
+	for (auto &bardef: bars) {
+		hashCombine(result, bardef.GetHash());
+	}
+	return result;
+}
+
+std::size_t std::hash<C4EnergyBarsDef>::operator()(const C4EnergyBarsDef &value) const noexcept
+{
+	return value.GetHash();
+}
+
+std::shared_ptr<C4FacetExID> C4EnergyBarsUniquifier::GetFacet(C4EnergyBarsDef::Gfxs &gfxs, const char *gfx)
+{
+	std::string key(gfx);
+
+	try {
+		auto facet = graphics.at(key).lock();
+		if(facet) {
+			return facet;
+		}
+	} catch (const std::out_of_range &) {
+	}
+
+	// facet needs to be loaded
+	int32_t amount = 0;
+	int32_t scale  = 100;
+	std::string file;
+	try {
+		auto gfx = gfxs.at(key);
+		amount = gfx.amount;
+		scale = gfx.scale;
+		file = gfx.file;
+	} catch (const std::out_of_range &) {
+		LogF("DefineEnergyBars: Missing key '%s' in graphics definition", key.c_str());
+		return nullptr;
+	}
+
+	// TODO FIXME why wasn't this already called before we get here?
+	Game.GraphicsResource.RegisterMainGroups();
+
+	auto facet = std::shared_ptr<C4FacetExID>(new C4FacetExID(), [=](C4FacetExID* facet){graphics.erase(key);});
+	bool success = Game.GraphicsResource.LoadFile(*facet, file.c_str(), Game.GraphicsResource.Files);
+	if(!success) {LogF("could not load custom energy bar graphic '%s'", file.c_str()); return nullptr;}
+
+	int32_t bar_wdt = facet->Surface->Wdt / (amount * 2);
+	int32_t bar_hgt = facet->Surface->Hgt / 3;
+	bar_wdt = (bar_wdt*100) / scale;
+	bar_hgt = (bar_hgt*100) / scale;
+	facet->Set(facet->Surface, 0, 0, bar_wdt, bar_hgt);
+
+	graphics.emplace(key, std::weak_ptr<C4FacetExID>(facet));
+	return facet;
+}
+
+std::shared_ptr<C4EnergyBarsDef> C4EnergyBarsUniquifier::UniqueifyDefinition(C4EnergyBarsDef *definition)
+{
+	// TODO somehow change *definition to unique_ptr?
+	// currently definition always gets owned by a shared_ptr, that either ends up being the canonical shared_ptr
+	// or goes out of scope deleting the definition
+
+	// the weak ptr remembers the custom deleter
+	auto shared = std::shared_ptr<C4EnergyBarsDef>(definition, [=](C4EnergyBarsDef* def){definitions.erase(*def);});
+	auto it_success = definitions.emplace(*definition, std::weak_ptr<C4EnergyBarsDef>(shared));
+	if (!it_success.second) {
+		// definition already existed, we have to override shared so that it is linked to
+		// the same control block that was used to create the first shared ptr
+		shared = it_success.first->second.lock();
+	}
+	return shared;
+}
+
+std::shared_ptr<C4EnergyBars> C4EnergyBarsUniquifier::DefineEnergyBars(C4ValueHash* graphics, C4ValueArray *definition)
+{
+	if (!graphics) {
+		LogF("DefineEnergyBars Parameter 1 map expected but got: %s", C4Value(graphics).GetDataString().getData());
+		return nullptr;
+	}
+
+	if (!definition) {
+		LogF("DefineEnergyBars Parameter 2 array expected but got: %s", C4Value(definition).GetDataString().getData());
+		return nullptr;
+	}
```

Comment:
The messages seem like it should be a script error, not just a log message.
In that case, it might be better suited in the script functions and the precondition of non-null arguments could be manifested by asserts or references.
If the condition is true, isn’t the value always `nil`? Would make the GetDataString() completely unnecessary.
Also, could be simply deduplicated into a local lambda, especially when throwing for error, instead of logging.

Reply:
I agree, sounds good, will look into it. Taking a break for now.

Reply:
these messages in particular are gone because graphics and definition are now references and when one of them is nil (on the caller side) default bars are restored.
Changed logs to script exceptions, local lambdas might still be useful to shorten code.